csrc/layernorm_kernels.cu

```diff
@@ -23,86 +23,88 @@
-  }
+  const int64_t tx = threadIdx.x;
+  const int64_t bx = blockIdx.x;
+  const int64_t num_threads = blockDim.x;
 
-  __device__ vec8_t operator*(const float& scale) const {
-    return vec8_t(x * scale, y * scale, z * scale, w * scale, u * scale,
-                  v * scale, s * scale, t * scale);
-  }
+  auto* __restrict__ out_v = reinterpret_cast<_f16Vec<scalar_t, width>*>(out);
+  auto* __restrict__ input_v =
+      reinterpret_cast<const _f16Vec<scalar_t, width>*>(
+          input + bx * static_cast<int64_t>(hidden_size));
+  auto* __restrict__ weight_v =
+      reinterpret_cast<const _f16Vec<scalar_t, width>*>(weight);
 
-  __device__ vec8_t operator+(const vec8_t& other) const {
-    return vec8_t(x + other.x, y + other.y, z + other.z, w + other.w,
-                  u + other.u, v + other.v, s + other.s, t + other.t);
+  // Compute variance. Be careful, hidden_size should multiple of 4.
+  for (int idx = tx; idx < vec_hidden_size; idx += num_threads) {
+    _f16Vec<scalar_t, width> temp = input_v[idx];
+    v8_variance_sum += temp.sum_squares();
   }
 
-  __device__ void operator+=(const vec8_t& other) {
-    x += other.x;
-    y += other.y;
-    z += other.z;
-    w += other.w;
-    u += other.u;
-    v += other.v;
-    s += other.s;
-    t += other.t;
-  }
+  using BlockReduce = cub::BlockReduce<float, 1024>;
+  __shared__ typename BlockReduce::TempStorage reduceStore;
 
-  __device__ scalar_t sum() const { return x + y + z + w + u + v + s + t; }
-};
+  float variance =
+      BlockReduce(reduceStore).Reduce(v8_variance_sum, cub::Sum{}, num_threads);
 
-#ifdef __HIP__MI300_MI250__
+  if (threadIdx.x == 0) {
+    s_variance = rsqrtf(variance / hidden_size + epsilon);
+  }
+  __syncthreads();
 
-// TODO(woosuk): Further optimize this kernel.
-template <typename scalar_t>
-__global__ void rms_norm_kernel(
-    scalar_t* __restrict__ out,           // [..., hidden_size]
-    const scalar_t* __restrict__ input,   // [..., hidden_size]
-    const scalar_t* __restrict__ weight,  // [hidden_size]
-    const float epsilon, const int num_tokens, const int hidden_size) {
-  __shared__ float s_variance;
+  variance = s_variance;
 
-  vec8_t<scalar_t> v8_variance = {0, 0, 0, 0, 0, 0, 0, 0};
+  for (int idx = tx; idx < vec_hidden_size; idx += num_threads) {
+    _f16Vec<scalar_t, width> temp = input_v[idx];
+    temp *= variance;
+    temp *= weight_v[idx];
+    out_v[bx * static_cast<int64_t>(vec_hidden_size) + idx] = temp;
+  }
+}
 
-  vec8_t<scalar_t>* vectorized_out = reinterpret_cast<vec8_t<scalar_t>*>(out);
-  vec8_t<scalar_t> const* vectorized_in =
-      reinterpret_cast<vec8_t<scalar_t> const*>(input);
-  vec8_t<scalar_t> const* vectorized_weight =
-      reinterpret_cast<vec8_t<scalar_t> const*>(weight);
-  const int vec_hidden_size = hidden_size >> 3;
+template <typename scalar_t, int width>
+__global__ std::enable_if_t<(width == 0) || !_typeConvert<scalar_t>::exists>
```

Comment:
Since this is effectively a copy paste from the mom MI version of the kernel, could we maybe change the define __HIP__MI300_MI250__ to be either 0 or 1, instead of defined or not defined, and add it to the std::enable_if to avoid duplication?